### ISSUE TITLE
Better timestamp handling

### ISSUE THREAD

**Issue by nagix** [drewnoakes/metadata-extractor#147]
The following fields don't contain the time zone information, and are stored into directories as strings. This seems reasonable because applications will mange how to map those timestamps to the specific time zone by using Directory.getDate(int tagType, TimeZone timeZone).

- ExifIFD0Directory.TAG_DATETIME
- ExifSubIFDDirectory.TAG_DATETIME_ORIGINAL
- ExifSubIFDDirectory.TAG_DATETIME_DIGITIZED

On the other hand, the following fields don't contain the time zone information either, but are interpreted as UTC time and stored into directories as Date objects. Applications are not able to directly map them to other time zones even if they are specified in Directory.getDate(int tagType, TimeZone timeZone).

- IccDirectory.TAG_PROFILE_DATETIME
- PngDirectory.TAG_LAST_MODIFICATION_TIME
- XmpDirectory.TAG_DATETIME_ORIGINAL
- XmpDirectory.TAG_DATETIME_DIGITIZED

There is another issue with the following fields. IPTC data contains the time zone information, but it is included only in the time fields. So, the time fields and its corresponding date fields need to be handled together.

- IptcDirectory.TAG_DATE_SENT and IptcDirectory.TAG_TIME_SENT
- IptcDirectory.TAG_RELEASE_DATE and IptcDirectory.TAG_RELEASE_TIME
- IptcDirectory.TAG_EXPIRATION_DATE and IptcDirectory.TAG_EXPIRATION_TIME
- IptcDirectory.TAG_DATE_CREATED and IptcDirectory.TAG_TIME_CREATED
- IptcDirectory.TAG_DIGITAL_DATE_CREATED and IptcDirectory.TAG_DIGITAL_TIME_CREATED 
- IptcDirectory.TAG_REFERENCE_DATE

The following fields represent a UTC timestamp, which are also need to be handled together.

- GpsDirectory.TAG_DATE_STAMP and GpsDirectory.TAG_TIME_STAMP

Also, the following fields hold the sub-second values for the timestamps, and should be reflected in the corresponding Date objects.

- ExifSubIFDDirectory.TAG_SUBSECOND_TIME
- ExifSubIFDDirectory.TAG_SUBSECOND_TIME_ORIGINAL
- ExifSubIFDDirectory.TAG_SUBSECOND_TIME_DIGITIZED

With that, my suggestion is as follows:

- Store a timestamp value into a directory as a String object if it has no time zone data
- Store a timestamp value into a directory as a Date object if it has time zone data
- For separate date/time fields, store a date value and a time value separately as a String, but convert them to a single Date object and store it to the both fields once both values became available
- For sub-second fields, reformat String objects for timestamps if they exists

**Comment by drewnoakes** [comment]
This is a great analysis.

In the case where there's a 1:1 mapping, the value stored in the directory should have as much information as possible, and no more :) Ideally parsing should happen when values are read from the directory, not when they are written in. This does not happen everywhere with the library, and this inconsistency should be addressed (See also drewnoakes/metadata-extractor-dotnet#34 where string encodings are decided during parsing, which is inflexible).

In the case where a single logical value (i.e. a DateTime) is derived from more than one physical tag, it's less clear to me what the best solution is. I'd like to keep the contents of a Directory as close to the representation in the file as possible (as above, and this will be useful when writing metadata back to files in future). One option is to add `get*` methods to `*Directory` classes that read from multiple underlying tags. The only downside to these is that they don't appear in current enumerations of all tags. Perhaps the concept of such 'virtual' tags could be added.

---

It seems to me that there are two concrete tasks here.

Firstly, modifying...

- IccDirectory.TAG_PROFILE_DATETIME
- PngDirectory.TAG_LAST_MODIFICATION_TIME
- XmpDirectory.TAG_DATETIME_ORIGINAL
- XmpDirectory.TAG_DATETIME_DIGITIZED

...to store strings is clearly a good thing and enables specification of time zone by the reader.

Secondly, adding `get*` methods to `IptcDirectory`, `GpsDirectory` and `ExifSubIFDDirectory` that draw from underlying tags to produce complete `Date` objects.

An optional third task is to consider surfacing these virtual tags in tag enumeration of directories.

How does that sound?

**Comment by nagix** [comment]
I agree with the idea that the value stored in the directory should be as close to that in the file as possible, and having additional `get*` methods as `GpsDirectory.getGeoLocation()` is actually doing. 

Currently, `IptcReader` stores the values for `TAG_RELEASE_DATE`, `TAG_DATE_CREATED` and `TAG_DIGITAL_DATE_CREATED` as `Date` with time value 00:00:00 in the default time zone (not even UTC). This is problematic, so I will make a change so that only the date part of them are stored as `String` and accessible via `getString` or new `get*` methods.

For the virtual tags, I like that idea. I think it should be addressed in a separate issue.

**Comment by drewnoakes** [comment]
Finally found some time to go through these. Great analysis, and great PRs. Thank you on behalf of the whole community.

In .NET (which I use more than Java these days) there's a `DateTime` class which allows for dates in an unspecified time zone. This doesn't appear to be something Java offers out of the box (Joda time seems an option) but it does seem like it'd be a good fit for many of these cases.

At any rate the solution here is an improvement. The only downside I can imagine people might notice is an inconsistency between the descriptions of date/time tags across different metadata types. This can be dealt with in descriptors though.

Thanks again!

**Comment by drewnoakes** [comment]
Regarding virtual tags, the idea is currently being tracked in #10.

I'll leave this issue open for you to close if you're happy with the current state of things.

**Comment by nagix** [comment]
I was also thinking about the use of a class like what you mentioned (`DateTime` in .NET or `org.joda.time.LocalDateTime`), and it was `com.adobe.xmp.XMPDateTime` in my case. It keeps information about whether the date has a specific time zone or not. Either one seems to cover these cases, and the code will be cleaner. I can create another issue and work on it. Do you think it is a good idea to leverage an existing class such as `org.joda.time.LocalDateTime` and `com.adobe.xmp.XMPDateTime`, or to create `com.drew.lang.DateTime` with the similar functions?

**Comment by drewnoakes** [comment]
I haven't used Joda before. One perk may be that it has a .NET port. Custom
class means no dependencies, which is always a good thing, so long as it's
easy to use. Joda has a lot of stuff we don't need, but it's well tested
and widely used.

What do you think makes the most sense?

Also, I've ported your changes to the .NET project but haven't pushed yet.
In .NET I think DateTime and DateTimeOffset might be sufficient.

**Comment by nagix** [comment]
As you say, Joda seems to be too much. I like a custom class which has no dependencies. It should be simple, with several setter/getter methods, a parser method and converter methods to `Date`, `String` and several custom string formats like EXIF, IPTC and XMP (ISO 8601) date format.

Another question is whether it should be exposed to applications through API (e.g. creating `Directory.getDateTime()` method), or keep it for internal use. Maybe some applications want to use it directly.
